src/main/java/org/candlepin/subscriptions/marketplace/MarketplacePayloadMapper.java

```diff
@@ -0,0 +1,199 @@
+     * "_ANY".  We only want to report on hourly metrics, so the Granularity should be HOURLY.
+     *
+     * @param snapshot tally snapshot
+     * @return eligibility status
+     */
+    protected boolean isSnapshotPAYGEligible(TallySnapshot snapshot) {
+        String productId = snapshot.getProductId();
+        var applicableProducts = List.of("OpenShift-metrics", "OpenShift-dedicated-metrics");
+        boolean isApplicableProduct = applicableProducts.contains(productId);
+
+        boolean isHourlyGranularity = Objects.nonNull(snapshot.getGranularity()) &&
+            Objects.equals(TallySnapshot.Granularity.HOURLY, snapshot.getGranularity());
+
+        boolean isSpecificUsage = Objects.nonNull(snapshot.getUsage()) &&
+            !Objects.equals(snapshot.getUsage(), TallySnapshot.Usage.ANY);
+
+        boolean isSpecificServiceLevel =
+            Objects.nonNull(snapshot.getSla()) && !Objects.equals(snapshot.getSla(), TallySnapshot.Sla.ANY);
+
+        boolean isSnapshotPAYGEligible =
+            isHourlyGranularity && isApplicableProduct && isSpecificUsage && isSpecificServiceLevel;
+
+        if (!isSnapshotPAYGEligible) {
+            log.debug("Snapshot not eligible for sending to RHM {}", snapshot);
+        }
+        return isSnapshotPAYGEligible;
+    }
+
+    /**
+     * UsageRequest objects are made up of a list of UsageEvents.
+     *
+     * @param tallySummary TallySummary
+     * @return List<UsageEvent>
+     */
+    protected List<UsageEvent> produceUsageEvents(TallySummary tallySummary) {
+        if (Objects.isNull(tallySummary.getTallySnapshots())) {
+            tallySummary.setTallySnapshots(new ArrayList<>());
+        }
+
+        return tallySummary.getTallySnapshots().stream().filter(this::isSnapshotPAYGEligible)
+            .map(snapshot -> {
+                String productId = snapshot.getProductId();
+                OffsetDateTime snapshotDate = snapshot.getSnapshotDate();
+                String eventId = snapshot.getId().toString();
+
+                // call MarketplaceIdProvider.findSubscriptionId once available
+                UsageCalculation.Key usageKey = new UsageCalculation.Key(productId,
+                    ServiceLevel.fromString(snapshot.getSla().toString()),
+                    Usage.fromString(snapshot.getUsage().toString()));
+
+
+                /*
+                This will need to be updated if we expand the criteria defined in the
+                isSnapshotPAYGEligible method to allow for Granularities other than HOURLY
+                 */
+                OffsetDateTime startDate = snapshotDate.minus(Duration.of(1, ChronoUnit.HOURS));
+
+                long start = startDate.toEpochSecond();
+                long end = snapshotDate.toEpochSecond();
+
+                var subscriptionId = findSubscriptionId(tallySummary.getAccountNumber(), usageKey,
+                    startDate, snapshotDate).orElseThrow();
```

Comment:
How does the thrown exception work in the context of a lambda?  I'm assuming/guessing it just brings everything to a halt.  Is that what we want?  Might be worth throwing a semantically meaningful exception.

Reply:
Good point.  Refactoring to be a for each loop that I can log and `continue` out of.